### ISSUE TITLE
Are there plans to support publishing through the "Publisher Portal"?

### ISSUE THREAD

**Issue by NotMyFault** [gradle-nexus/publish-plugin#330]
Hey,

sonatype shut down registering new namespaces for the OSSRH earlier this year. New registrations are now done through the [publisher portal](https://central.sonatype.com/publishing).
If I read the [documentation](https://central.sonatype.org/publish-ea/publish-ea-guide/) right, there is only a [maven plugin](https://central.sonatype.org/publish/publish-portal-maven/) at the moment.

It would be awesome if this gradle plugin would support it too, for upcoming namespaces people register.

**Comment by szpak** [comment]
Briefly reading the documentation, I'm not sure if/when Sonatype plans to retire releasing (new) projects via OSSRH. In addition, I wasn't able to find any developer (API) documentation for releasing (and even their Maven plugin doesn't seem to be available on GitHub, or anywhere else with the source code - or maybe, I just wasn't able to find it).

@NotMyFault Could you ask the Sonatype guys about that?

**Comment by NotMyFault** [comment]
> I'm not sure if/when Sonatype plans to retire releasing (new) projects via OSSRH

Afaik, existing projects can continue to use the "old" workflow to publishing to maven central through OSSRH (for now), but new registrations have to use https://central.sonatype.com/publishing.
Large parts of the pre-existing workflow are already gone, e.g. the Jira instance to request and update release permissions now points to the central portal: https://issues.sonatype.org/

> I wasn't able to find any developer (API) documentation

There appears to be a bit at https://central.sonatype.org/publish/publish-portal-api/

> Nevertheless, even reverse engineering existing code could be time consuming - the API documentation would be much better

Agreed, I'll happily ask if the source code lives somewhere else. Although there appear to exist a few gradle plugins for the central portal already: https://central.sonatype.org/publish/publish-portal-gradle/
But none of them suffice for my workflow, like yours :(

**Comment by szpak** [comment]
Thanks for the extra information.

We would probably need to analyze how similar the workflow is and much would it be simple to put it into this plugin to handle it in the other way round. Also, it seems to be still "Publisher Early Access", so the things might change (although the new projects/namespaces seem to need to use the new approach, unless [manually migrated to OSSRH](https://central.sonatype.org/register/central-portal/#switching-to-ossrh-during-portal-early-access) :-/ ).

**Comment by NotMyFault** [comment]
> Also, it seems to be still "Publisher Early Access", so the things might change (although the new projects/namespaces seem to need to use the new approach, unless [manually migrated to OSSRH](https://central.sonatype.org/register/central-portal/#switching-to-ossrh-during-portal-early-access) :-/ ).

Yeah, the time to retire the old service was really short. According [to the setup docs](https://central.sonatype.org/publish-ea/publish-ea-guide/#managing-your-credentials), all new namespace registrations only happen through the new portal.

**Comment by NotMyFault** [comment]
> > Also, it seems to be still "Publisher Early Access", so the things might change (although the new projects/namespaces seem to need to use the new approach, unless [manually migrated to OSSRH](https://central.sonatype.org/register/central-portal/#switching-to-ossrh-during-portal-early-access) :-/ ).
> 
> Yeah, the time to retire the old service was really short. According [to the setup docs](https://central.sonatype.org/publish-ea/publish-ea-guide/#managing-your-credentials), all new namespace registrations only happen through the new portal.

After reaching out, they told me there are plans to open source the plugin, but no date when this will hapen. I could take a look at the -sources jar, if I want :/

**Comment by bric3** [comment]
Sonatype docs now points to use the jreleaser instead : https://central.sonatype.org/publish/publish-portal-gradle/

**Comment by szpak** [comment]
Thanks for pointing it out Bruce, I didn't know that JReleaser can publish to Maven Central :thinking: . We will have to take a look at their implementation and how much their functionality overlaps with this plugin.

**Comment by bric3** [comment]
I just discovered it today ;)

JReleaser also support publishing the release to github, which could be practical to replace shipkit as well.

**Comment by szpak** [comment]
Btw, anyone tried to use the "emulation API" with this plugin?
https://central.sonatype.org/publish/publish-portal-ossrh-staging-api/

**Comment by wborn** [comment]
> Btw, anyone tried to use the "emulation API" with this plugin?

I did a bit of testing with it today on a namespace registered in the Central Portal and so far it seems to work well. The plugin properly created a staging repository and closed it when using `publishToSonatype closeSonatypeStagingRepository`. After that it showed up as a [deployment](https://central.sonatype.com/publishing) in the UI that can be dropped or published.

I haven't tested using `closeAndReleaseSonatypeStagingRepository` (which I normally use) yet but it looks promising. 

So the only changes you need to make is:

1. Generate/configure a new user token (https://central.sonatype.com/account)
2. Update the URLs in Gradle:
  i. Snaphots: https://central.sonatype.com/repository/maven-snapshots/
  ii. Releases: https://ossrh-staging-api.central.sonatype.com/service/local/

**Comment by szpak** [comment]
Thanks for checking @wborn!

**Comment by chadlwilson** [comment]
I can confirm that this plugin seems to work fine with @wborn's URL change to use the compatibility API, for a migrated namespace (`cd.go`). (using `publishToSonatype` with `closeAndReleaseSonatypeStagingRepository`)

![Image](https://github.com/user-attachments/assets/836749f9-edd4-4787-9216-26626c47e45e)

**Comment by tomasbjerre** [comment]
I can also confirm that @wborn comment (https://github.com/gradle-nexus/publish-plugin/issues/330#issuecomment-2866996208) is all you need to do to migrate. I think it should be added to the README.

**Comment by wborn** [comment]
> I think it should be added to the README. 

There's PR #377 to do this.

**Comment by NetPumi2** [comment]
For publishing to the Maven portal, I am still using this plugin (very old version and I can not update), but just for making my module and then I wrap it as a zip file and send it to the API 
```   curl --request POST \
  --verbose \
  --header 'Authorization: Bearer ${$MAVEN_PORTAL_PUBLISHING_BEARER_TOKEN}' \
  --form bundle=@${zipPath} \
  'https://central.sonatype.com/api/v1/publisher/upload?publishingType=USER_MANAGED'``` and it is working just fine

**Comment by Ayfri** [comment]
Any update?

**Comment by TWiStErRob** [comment]
> anyone tried to use the "emulation API" with this plugin?

It worked well for me for both snapshot and central publishing. You can see the required changes here: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/951

**Comment by Vampire** [comment]
Using the compatibility API works, but it might be beneficial if the new way would work too, because I heard it is ultimately faster. For a single-artifact project that might not be that relevant, but for bigger projects it might be.

**Comment by wborn** [comment]
It could be that the because of the compatibility API slowness our releases often fail (https://github.com/openremote/openremote/issues/2523) nowadays with:

```
Execution failed for task ':closeSonatypeStagingRepository'.
> Failed to close staging repository, server at https://ossrh-staging-api.central.sonatype.com/service/local/ responded with status code 400, body: Failed to process request: No objects found in the repository
```

We use `gradle publishToSonatype closeAndReleaseSonatypeStagingRepository` 

It doesn't look like the retry mechanism in the plugin can properly handle this. That's because there doesn't seem to be staging repository yet so it cannot transition it.
